pkg/reporter/cyclonedx.go

```diff
@@ -0,0 +1,302 @@
+	}
+
+	r.recordDependencies(pkg)
+	r.recordVulnerabilities(pkg)
+
+	*r.bom.Components = append(*r.bom.Components, component)
+}
+
+func (r *cycloneDXReporter) resolvePackageLicenses(pkg *models.Package) []cdx.LicenseChoice {
+	licenses := []cdx.LicenseChoice{}
+
+	if pkg.Insights == nil || pkg.Insights.Licenses == nil {
+		return licenses
+	}
+	for _, license := range *pkg.Insights.Licenses {
+		licenseChoice := cdx.LicenseChoice{
+			License: &cdx.License{
+				Name: string(license),
+				ID:   string(license),
+			},
+		}
+		if _, validSpdxLicense := data.SPDXLicenseMap[string(license)]; validSpdxLicense {
+			// @TODO - this is taken from https://github.com/CycloneDX/cdxgen/blob/c32b2b64174c190490e34e18fe7c9a3f7bb4904e/lib/helpers/utils.js#L814
+			// However, some licenses don't point to a valid URL
+			licenseChoice.License.URL = "https://opensource.org/licenses/" + string(license)
+		}
+		licenses = append(licenses, licenseChoice)
+	}
+
+	return licenses
+}
+
+func (r *cycloneDXReporter) recordDependencies(pkg *models.Package) {
+	pkgPurl := pkg.GetPackageUrl()
+
+	deps, err := pkg.GetDependencies()
+	if err != nil {
+		logger.Errorf("[CycloneDX Reporter]: Failed to get dependencies for package %s: %v", pkgPurl, err)
+		return
+	}
+
+	depsPurls := []string{}
+	for _, dep := range deps {
+		depsPurls = append(depsPurls, dep.GetPackageUrl())
+	}
+
+	*r.bom.Dependencies = append(*r.bom.Dependencies, cdx.Dependency{
+		Ref:          pkgPurl,
+		Dependencies: &depsPurls,
+	})
+}
+
+func (r *cycloneDXReporter) recordVulnerabilities(pkg *models.Package) {
+	if pkg.Insights == nil || pkg.Insights.Vulnerabilities == nil {
+		return
+	}
+
+	pkgPurl := pkg.GetPackageUrl()
+
+	for _, vuln := range *pkg.Insights.Vulnerabilities {
+		ratings := []cdx.VulnerabilityRating{}
+		for _, severity := range utils.SafelyGetValue(vuln.Severities) {
+			rating := cdx.VulnerabilityRating{
+				Method:   cdx.ScoringMethod(utils.SafelyGetValue(severity.Type)),
+				Severity: cdx.Severity(utils.SafelyGetValue(severity.Risk)),
+				Vector:   utils.SafelyGetValue(severity.Score),
+			}
+			ratings = append(ratings, rating)
+		}
+
+		*r.bom.Vulnerabilities = append(*r.bom.Vulnerabilities, cdx.Vulnerability{
+			ID:          utils.SafelyGetValue(vuln.Id),
+			BOMRef:      strings.Join([]string{utils.SafelyGetValue(vuln.Id), pkgPurl}, "/"), // This format of Vulnerability BOMRef is used by cdxgen
+			Description: utils.SafelyGetValue(vuln.Summary),
+			Ratings:     &ratings,
+			Affects: commonUtils.PtrTo([]cdx.Affects{
+				{
+					Ref: pkgPurl,
+				},
+			}),
+		})
+	}
+}
+
+func (r *cycloneDXReporter) AddAnalyzerEvent(event *analyzer.AnalyzerEvent) {
+}
+
+func (r *cycloneDXReporter) AddPolicyEvent(event *policy.PolicyEvent) {
+	// @TODO - How to handle policy events
+}
+
+func (r *cycloneDXReporter) finaliseBom() {
+	bomGenerationTime := time.Now().UTC()
+
+	r.bom.Metadata.Timestamp = bomGenerationTime.Format(time.RFC3339)
+
+	annotation := (*r.bom.Annotations)[0]
```

Comment:
This seems weird. Especially since we are overwriting `r.bom.Annotations` in line number 278 anyway

Reply:
creating annotations during finalise phase to prevent this